### PR TITLE
feat(commands): Implement argument handling for custom commands via a prompt pipeline

### PR DESCRIPTION
**Pull request by abhipatel12**
## TLDR

This PR introduces argument handling for custom TOML commands, enabling two powerful new modes: **Shorthand Injection** (via `{{args}}`) and **Default Prompt Processing**.

The core of this change is a new, extensible **Prompt Processing Pipeline** architecture. This pipeline is the key enabler for future command enhancements such as shell command processing.

**Key areas for review:**
*   The new `prompt-processors/` directory, containing the pipeline interface (`types.ts`) and the new argument processors (`argumentProcessor.ts`).
*   The refactored `FileCommandLoader.ts`, which now acts as a "pipeline assembler" by building a list of processors for each command's action.
*   The new unit tests in `argumentProcessor.test.ts` and the updated integration tests in `FileCommandLoader.test.ts`.
*   The updated user documentation in `docs/cli/commands.md`.

## Dive Deeper

The implementation is centered on a **Prompt Processing Pipeline**.

### The Architectural Decision: A Prompt Pipeline

Instead of creating a monolithic `action` for each file-based command, this change introduces a pipeline of small, single-purpose "processors." When a command is loaded, the `FileCommandLoader` inspects its `prompt` (and eventually other  keys in the `.toml`) and assembles the correct pipeline. The command's `action` then simply executes its assigned processors in sequence.

This architecture was chosen for three key reasons:

1.  **Composability:** This is the primary benefit. Our features are not mutually exclusive. A future command might need shell execution (`!git diff`) *and* argument injection (`{{args}}`). The pipeline allows us to simply add a `ShellProcessor` and an `ArgumentProcessor` to the chain, and they work together seamlessly. This avoids complex factories and keeps the code clean and maintainable.
2.  **Extensibility:** Adding new prompt features (e.g., file injection with `@path`) becomes a matter of creating a new, isolated processor and adding it to the assembly logic in `FileCommandLoader`. The core engine remains untouched.
3.  **Testability & Separation of Concerns:** Each processor is a pure, testable unit. The `ShorthandArgumentProcessor` knows nothing about the file system; it only knows how to replace a string. This clear separation makes the system dramatically easier to reason about and test.

This PR establishes this pipeline and provides its first two processors for handling arguments, which manifest as two distinct user-facing features.

### The User-Facing Features

1.  **Shorthand Injection (`{{args}}`):** If a command's `prompt` contains `{{args}}`, the pipeline uses the `ShorthandArgumentProcessor` to inject all text following the command directly into that placeholder. This is for simple, deterministic templating.

2.  **Default Prompt Processing:** If `{{args}}` is absent, the pipeline uses the `DefaultArgumentProcessor`. This processor appends the user's *entire raw input* (e.g., `/changelog 1.2.0 added "a feature"`) to the bottom of the base prompt. This provides the model with full context to act as a "mini-API" interpreter, enabling far more complex and agentic commands.

## Reviewer Test Plan

To validate these changes, please pull down the branch and perform the following manual tests after verifying that all automated tests pass with `npm run preflight`.

#### 1. Test Shorthand Injection (`{{args}}`)
1.  Create `~/.gemini/commands/shorthand.toml` with the following content:

    ```toml
    description = "A test for shorthand {{args}} injection."
    prompt = "The user wants to know about the following topic: {{args}}. Please provide a detailed explanation."
    ```
3.  Run `npm start` and execute: `/shorthand Why is pluto not a planet?`
4.  **Expected:** The model should respond as if it received the prompt: `The user wants to know about the following topic: Why is pluto not a planet?. Please provide a detailed explanation.` The final prompt sent to the LLM is constructed correctly.

#### 2. Test Default Interpretation
1.  In a test project directory, create `.gemini/commands/git/feature.toml`:

    ```toml
    # In: <project>/.gemini/commands/git/feature.toml

    description = "Creates a new feature branch from a ticket ID and description."
    prompt = """
    # Task: Create Git Feature Branch

    You are an expert at using Git. The user has invoked a command to create a new feature branch.

    ## Expected Format
    The command must follow this format: `/git:feature <ticket-id> <description>`

    ## Your Task
    1. Parse the `<ticket-id>` and `<description>` from the user's input.
    3. Sanitize the description to be URL-friendly (lowercase, use hyphens for spaces).
    5. Construct a branch name in the format: `feature/<ticket-id>-<sanitized-description>`
    6. Use the `shell` tool to execute the `git checkout -b` command with the constructed branch name.
    """
    ```
2.  Run `npm start` from that project directory and execute the command:
    `/git:feature 123 Add new login button`
3.  **Expected:**
    *   The CLI should display a `shell` tool call being executed.
    *   The command being run should be `git checkout -b feature/123-add-new-login-button` (or a close, valid equivalent).
    *   After the tool runs, you can verify success by running `!git status` in the CLI. The output should confirm you are now on the newly created branch.

#### 3. Test Argument-less Default Processing Command (Edge Case)
1.  Using the same file from step 2, execute the command with no arguments: `/git:feature`
2.  **Expected:** The model should will not receive the raw input and may ask for clarification, referencing the `Expected Format`, should respond that it's missing the ticket ID and description. This proves that the context and parsing instructions are passed correctly.

#### 4. Verify Documentation
1.  Open the `docs/cli/commands.md` file in your editor or a Markdown previewer.
2.  **Expected:** Verify that the new `#### Handling Arguments` section is present and renders correctly, with clear explanations and examples for both the `{{args}}` and model-led methods.
## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✅  | ❓  | ❓  |
| npx      | ❓  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ✅  | -   | -   |

## Linked issues / bugs

This PR makes progress on #3789

### CODE REVIEW

File: packages/cli/src/services/FileCommandLoader.ts [hunk 1]
```diff
@@ -156,16 +166,41 @@ export class FileCommandLoader implements ICommandLoader {
       .map((segment) => segment.replaceAll(':', '_'))
       .join(':');
 
+    const processors: IPromptProcessor[] = [];
+
+    // The presence of '{{args}}' is the switch that determines the behavior.
+    if (validDef.prompt.includes('{{args}}')) {
```
[the remark this answers is not in the document]
Comment (abhipatel12) [reply packages/cli/src/services/FileCommandLoader.ts]: Done

File: packages/cli/src/services/prompt-processors/argumentProcessor.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * Replaces all instances of `{{args}}` in a prompt with the user-provided
+ * argument string.
+ */
+export class ShorthandArgumentProcessor implements IPromptProcessor {
+  async process(
+    prompt: string,
+    args: string,
+    _fullCommand: string,
+  ): Promise<string> {
+    return prompt.replaceAll('{{args}}', args);
```
[the remark this answers is not in the document]
Comment (abhipatel12) [reply packages/cli/src/services/prompt-processors/argumentProcessor.ts]: Same fix

File: packages/cli/src/services/prompt-processors/types.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+   * @param prompt The current state of the prompt string.
+   * @param args The raw argument string from the user's command invocation.
+   * @param fullCommand The full command the user typed (e.g., /git:commit foo).
+   * @returns The transformed prompt string as a Promise.
+   *
+   * @todo Add `context: CommandContext` back when more complex processors
+   * that require access to services or UI are needed.
+   */
+  process(prompt: string, args: string, fullCommand: string): Promise<string>;
+}
```
[the remark this answers is not in the document]
Comment (abhipatel12) [reply packages/cli/src/services/prompt-processors/types.ts]: Same fix

File: packages/cli/src/services/FileCommandLoader.ts [hunk 4]
```diff
@@ -15,7 +15,17 @@ import {
...
+  CommandContext,
+  CommandKind,
+  SlashCommand,
+  SubmitPromptActionReturn,
+} from '../ui/commands/types.js';
+import {
+  ModelLedArgumentProcessor,
+  ShorthandArgumentProcessor,
+} from './prompt-processors/argumentProcessor.js';
+import { IPromptProcessor } from './prompt-processors/types.js';
```
[the remark this answers is not in the document]
Comment (abhipatel12) [reply packages/cli/src/services/FileCommandLoader.ts]: Same fix

File: docs/cli/commands.md [hunk 5]
```diff
@@ -129,6 +129,65 @@ Your command definition files must be written in the TOML format and use the `.t
...
+
+description = "Generates a fix for a given GitHub issue."
+prompt = "Please analyze the staged git changes and provide a code fix for the issue described here: {{args}}."
+```
+
+The model will receive the final prompt: `Please analyze the staged git changes and provide a code fix for the issue described here: "Button is misaligned on mobile".`
+
+##### 2. Model-led Interpretation
+
+If your `prompt` does **not** contain `{{args}}`, the CLI assumes the model is smart enough to parse the arguments itself. It will prepend the _entire command the user typed_ to the top of your prompt, separated by a Markdown horizontal rule (`---`), providing the model with the full context to execute.
```
Comment (NTaylorMullen) [inline docs/cli/commands.md]: the `---` separation piece is a bit arcane. What if we were to take the simple route and just append any content to the end? This way it feels more like a prompt where the template is there and any mutation to the invocation itself is seen as additional instructions after the primary template?

```
... The prompt ...
... whatever was passed ...
```

File: packages/cli/src/services/FileCommandLoader.ts [hunk 6]
```diff
@@ -156,16 +169,41 @@ export class FileCommandLoader implements ICommandLoader {
       .map((segment) => segment.replaceAll(':', '_'))
       .join(':');
 
+    const processors: IPromptProcessor[] = [];
+
+    // The presence of '{{args}}' is the switch that determines the behavior.
+    if (validDef.prompt.includes(SHORTHAND_ARGS_PLACEHOLDER)) {
+      processors.push(new ShorthandArgumentProcessor());
+    } else {
+      processors.push(new ModelLedArgumentProcessor());
```
Comment (NTaylorMullen) [inline packages/cli/src/services/FileCommandLoader.ts]: Instead of model lead prompts I'd actually just call it default prompt processor (if we take the append route)

File: packages/cli/src/services/FileCommandLoader.ts [hunk 7]
```diff
@@ -156,16 +169,41 @@ export class FileCommandLoader implements ICommandLoader {
...
-      action: async () => ({
-        type: 'submit_prompt',
-        content: validDef.prompt,
-      }),
+      action: async (
+        _context: CommandContext,
+        args: string,
+      ): Promise<SubmitPromptActionReturn> => {
+        let processedPrompt = validDef.prompt;
+        const fullCommand = `/${commandName} ${args}`.trim();
```
Comment (NTaylorMullen) [inline packages/cli/src/services/FileCommandLoader.ts]: This feels like something that should probably be on the `CommandContext` to capture the original user input

File: packages/cli/src/services/prompt-processors/argumentProcessor.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+ * Markdown horizontal rule. This provides a clean, structured context for
+ * the model to perform its own argument parsing based on the user's instructions.
+ */
+export class ModelLedArgumentProcessor implements IPromptProcessor {
+  async process(
+    prompt: string,
+    _args: string,
+    fullCommand: string,
+  ): Promise<string> {
+    return `${fullCommand}\n\n---\n\n${prompt}`;
```
Comment (NTaylorMullen) [inline packages/cli/src/services/prompt-processors/argumentProcessor.ts]: Have to say it's super interesting to show the model the full command, intention being to enable the command to auto-correct any faulty syntax.

File: packages/cli/src/services/prompt-processors/types.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+   *
+   * @param prompt The current state of the prompt string.
+   * @param args The raw argument string from the user's command invocation.
+   * @param fullCommand The full command the user typed (e.g., /git:commit foo).
+   * @returns The transformed prompt string as a Promise.
+   *
+   * @todo Add `context: CommandContext` back when more complex processors
+   * that require access to services or UI are needed.
+   */
+  process(prompt: string, args: string, fullCommand: string): Promise<string>;
```
Comment (NTaylorMullen) [inline packages/cli/src/services/prompt-processors/types.ts]: We should probably have a dedicated context object to make future iterations here easier

**Review by NTaylorMullen — APPROVED** [review]
🎉

File: docs/cli/commands.md (hunk 5 above)
Comment (abhipatel12) [reply docs/cli/commands.md]: Sounds good! Made changes to the doc and implementation.

File: packages/cli/src/services/prompt-processors/argumentProcessor.ts (hunk 8 above)
Comment (abhipatel12) [reply packages/cli/src/services/prompt-processors/argumentProcessor.ts]: This is cool b/c now the agent can clarify if needed or even be like "hey where's that arg". 

From the other comment, i've removed the `---` and just appended the full command to the bottom.

File: packages/cli/src/services/FileCommandLoader.ts (hunk 6 above)
Comment (abhipatel12) [reply packages/cli/src/services/FileCommandLoader.ts]: Done!

File: packages/cli/src/services/prompt-processors/types.ts (hunk 9 above)
Comment (abhipatel12) [reply packages/cli/src/services/prompt-processors/types.ts]: Moved this all into the `CommandContext`

File: packages/cli/src/services/FileCommandLoader.ts (hunk 7 above)
Comment (abhipatel12) [reply packages/cli/src/services/FileCommandLoader.ts]: Done!

**Comment by abhipatel12** [comment]
Thanks for the review! 

How do you feel about the `CommandContext` change here?

File: packages/cli/src/ui/commands/types.ts [hunk 10]
```diff
@@ -132,7 +141,7 @@ export interface SlashCommand {
   // The action to run. Optional for parent commands that only group sub-commands.
   action?: (
     context: CommandContext,
-    args: string,
+    args: string, // TODO: Remove args. CommandContext now contains the complete invocation.
```
Comment (NTaylorMullen) [inline packages/cli/src/ui/commands/types.ts:144]: agree should remove

File: packages/cli/src/ui/commands/types.ts [hunk 11]
```diff
@@ -14,6 +14,15 @@ import { SessionStatsState } from '../contexts/SessionContext.js';
...
 // Grouped dependencies for clarity and easier mocking
 export interface CommandContext {
+  // Invocation properties for when commands are called.
+  invocation?: {
+    /** The raw, untrimmed input string from the user. */
+    raw: string;
+    /** The primary name of the command that was matched. */
+    name: string;
+    /** The arguments string that follows the command name. */
+    args: string;
```
Comment (NTaylorMullen) [inline packages/cli/src/ui/commands/types.ts:24]: args we may want to be a map at some point

**Review by NTaylorMullen — APPROVED** [review]

File: packages/cli/src/ui/commands/types.ts (hunk 11 above)
Comment (abhipatel12) [reply packages/cli/src/ui/commands/types.ts:24]: Yep! That will be the next step. As soon as we have more complex arguments OR MCP prompt arguments, having this broken into a map will be great.

File: packages/cli/src/ui/commands/types.ts (hunk 10 above)
Comment (abhipatel12) [reply packages/cli/src/ui/commands/types.ts:144]: +1, left it after some of the work here slows just a little b/c it will touch every builtin command and didn't want to inflate this PR
